# Online Retail Sales & Customer Behavior Analysis

## 1.🧑‍💻 Introduction

This project focuses on analyzing an Online Retail Sales dataset to understand sales patterns, customer behavior, product performance, and other useful business trends.

The analysis follows a structured data analysis process, starting with understanding and investigating the dataset, followed by data cleaning, feature creation, exploratory data analysis, visualization, and insight generation.

The main objective is to transform raw transactional data into meaningful information and identify patterns that can support better business understanding and decision-making.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_excel("C:/Users/vikasa s y/Downloads/Online-Retail-Sales-Analysis/Data/Online Retail.xlsx")

## 2. Understanding The DataSet

In [3]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France


In [6]:
df.shape

(541909, 8)

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [8]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(5268)

In [11]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object

### 🧠 Dataset Understanding
- Our dataSet Contains **541909 Rows** And **8 Columns**
- The DataSet Conatins Some Missing Values Especially In **Description and CustomerID** Columns
- The Dataset Contains **5268** No Duplicate Rows
- In **Quantity and UnitPrice** Columns It Conatins Negitive Values That Require Further Investigation
- **Country And Description** Are The Categorical Data

---
---
## 3. Data Investigation

In [12]:
df["Quantity"].describe()

count    541909.000000
mean          9.552250
std         218.081158
min      -80995.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64

In [13]:
df["UnitPrice"].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

In [14]:
df["UnitPrice"].min()

np.float64(-11062.06)

In [15]:
df["UnitPrice"].max()

np.float64(38970.0)

In [16]:
df["Quantity"].max()

np.int64(80995)

In [17]:
df["Quantity"].min()

np.int64(-80995)

In [18]:
df[df["Quantity"] < 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [19]:
df[df["UnitPrice"] < 0].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [20]:
df[df["UnitPrice"] == 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


In [21]:
df[df["Quantity"] < 0]["InvoiceNo"].head(20)

141     C536379
154     C536383
235     C536391
236     C536391
237     C536391
238     C536391
239     C536391
240     C536391
241     C536391
939     C536506
1441    C536543
1442    C536543
1973    C536548
1974    C536548
1975    C536548
1976    C536548
1977    C536548
1978    C536548
1979    C536548
1980    C536548
Name: InvoiceNo, dtype: object

In [22]:
(df["UnitPrice"] < 0).sum()

np.int64(2)

In [23]:
(df["Quantity"] < 0).sum()

np.int64(10624)

In [24]:
(df["UnitPrice"] == 0).sum()

np.int64(2515)

In [25]:
df.loc[df["Quantity"].idxmax()]

InvoiceNo                           581483
StockCode                            23843
Description    PAPER CRAFT , LITTLE BIRDIE
Quantity                             80995
InvoiceDate            2011-12-09 09:15:00
UnitPrice                             2.08
CustomerID                         16446.0
Country                     United Kingdom
Name: 540421, dtype: object

In [26]:
df.loc[df["Quantity"].idxmin()]

InvoiceNo                          C581484
StockCode                            23843
Description    PAPER CRAFT , LITTLE BIRDIE
Quantity                            -80995
InvoiceDate            2011-12-09 09:27:00
UnitPrice                             2.08
CustomerID                         16446.0
Country                     United Kingdom
Name: 540422, dtype: object

### Investigation Findings
- `Description` Has **1454 Missing Values**
- `CustomerID` Have  **135080 Missing Values**
- `Quantity` Contains Negitive Values.
- Negative quantities are associated with invoice numbers beginning with `C`, indicating cancellation/return transactions.
- The dataset contains **10,624 negative UnitPrice values**, including records related to `Adjust bad debt`.
- There are **2,515 transactions with UnitPrice = 0**.
- An extreme quantity of **80,995** was found, along with a corresponding cancellation of **-80,995** for the same product.

### Initial Understanding

The unusual values are not automatically treated as errors. They represent different types of business transactions or special records. Therefore, cleaning decisions will be based on the purpose of the analysis rather than simply removing unusual values.

---
---
## 4.🧹 Data Cleaning

Based on the data investigation, several data-quality issues were identified, including missing values, duplicate records, cancellation transactions, unusual prices, and special transaction records.

The cleaning process will address these issues while preserving the original dataset. Each cleaning decision will be based on the purpose of the analysis and the meaning of the data.


In [27]:
df.shape

(541909, 8)

In [28]:
df[df.duplicated()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
587,536412,22273,FELTCRAFT DOLL MOLLY,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
589,536412,22749,FELTCRAFT PRINCESS CHARLOTTE DOLL,1,2010-12-01 11:49:00,3.75,17920.0,United Kingdom
594,536412,22141,CHRISTMAS CRAFT TREE TOP ANGEL,1,2010-12-01 11:49:00,2.10,17920.0,United Kingdom
598,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
600,536412,22569,FELTCRAFT CUSHION BUTTERFLY,2,2010-12-01 11:49:00,3.75,17920.0,United Kingdom


In [29]:
clean_df = df.drop_duplicates().copy()

In [30]:
clean_df.shape

(536641, 8)

### Duplicate Handling

- The dataset contained **5,268 exact duplicate rows**.
- These duplicate records were removed from the analysis dataset.
- The original dataset was preserved for reference.
- After removing duplicates, the dataset contains **536,641 rows**.

In [31]:
clean_df["CustomerID"].isna().sum()

np.int64(135037)

### CustomerID Missing Values

- A significant number of records have missing **CustomerID** values.
- These records are retained because `CustomerID` is not required for overall sales, product, country, or time-based analysis.
- Records with missing `CustomerID` will be excluded only when performing **customer-specific analysis**.

In [32]:
clean_df[clean_df["Description"].isna()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


---
---
### Handling Missing `Description`

In [33]:
clean_df = clean_df.dropna(subset=["Description"]).copy()

In [34]:
clean_df["Description"].isna().sum()

np.int64(0)

### Missing Description

- The dataset contained missing values in the `Description` column.
- Investigation showed that these records generally had a **UnitPrice of 0** and missing customer information.
- Since product-level analysis depends on identifying the product, records with missing descriptions were removed from the cleaned sales dataset.

---
---
### Handling Negative `Quantity`

In [35]:
Returns_df = df[df["Quantity"] < 0]

In [36]:
Returns_df.shape

(10624, 8)

In [37]:
sales_df = clean_df[clean_df["Quantity"] > 0].copy()

In [38]:
sales_df.shape

(525462, 8)

### Negative Quantity / Returns

- The original dataset contains **10,624 transactions with negative quantities**.
- These transactions are treated as return or cancellation records rather than invalid data.
- They are preserved separately in `returns_df`.
- A separate `sales_df` is created using positive quantities for normal sales analysis.

In [39]:
zero_qnty = df[df["Quantity"] == 0]

In [40]:
zero_qnty.shape

(0, 8)

In [41]:
zero_qnty.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


### Handling `UnitPrice`

In [42]:
negative_price = df[df["UnitPrice"] < 0]

In [43]:
negative_price.shape

(2, 8)

In [44]:
negative_price.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [45]:
clean_df[clean_df["UnitPrice"] <= 0].shape

(1058, 8)

In [46]:
clean_df[clean_df["UnitPrice"] <= 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,NaN,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,NaN,United Kingdom
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,NaN,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,NaN,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,NaN,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,NaN,United Kingdom


In [47]:
clean_df[clean_df["UnitPrice"] == 0].shape

(1056, 8)

In [48]:
clean_df[clean_df["UnitPrice"] == 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,NaN,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,NaN,United Kingdom
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,NaN,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,NaN,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,NaN,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,NaN,United Kingdom


### UnitPrice Analysis

- The `UnitPrice` column contains the price of each product per unit.
- The analysis identified **negative and zero UnitPrice values** that do not represent normal sales.
- Negative UnitPrice values were investigated and found to be associated with unusual financial adjustment records.
- Zero UnitPrice records represent transactions with no recorded sales price.
- Therefore, records with **UnitPrice <= 0** will be excluded from the normal sales analysis.
- Records with positive UnitPrice values will be retained for revenue and sales analysis.

---
---
### Extreme Quantity Values

- The dataset contains unusually large positive and negative quantity values.
- Investigation showed that the extreme values are associated with corresponding sales and cancellation transactions.
- Therefore, extreme quantities were not removed solely because they were large.
- Outlier treatment will only be considered when it is justified by the specific analysis being performed.

In [49]:
sales_df = clean_df[ (clean_df["Quantity"] > 0) & (clean_df["UnitPrice"] > 0) ].copy() 

In [50]:
sales_df.shape

(524878, 8)

In [51]:
sales_df.duplicated().sum()

np.int64(0)

In [52]:
sales_df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     132186
Country             0
dtype: int64

In [53]:
(sales_df["Quantity"] <= 0).sum()

np.int64(0)

In [54]:
(sales_df["UnitPrice"] <= 0).sum()

np.int64(0)

### Data Cleaning - Summary

The dataset was cleaned based on the findings from the initial data investigation.

- Removed **5,268 duplicate rows**.
- Retained records with missing `CustomerID` because they are still useful for overall sales analysis.
- Removed records with missing `Description` from the sales dataset.
- Separated **10,624 negative-quantity transactions** as returns/cancellations.
- Excluded transactions with `UnitPrice <= 0` from the normal sales dataset.
- Extreme quantity values were investigated and were not removed without sufficient justification.
- Created `sales_df` containing **524,878 valid sales transactions** for further analysis.

The original dataset was preserved, while the cleaned sales dataset will be used for feature creation and further exploratory analysis.

## 5. Feature Creation

### 💰 Feature 1 — Revenue Creation & Analysis

The dataset contains `Quantity` and `UnitPrice`, but it does not directly provide the total revenue generated by each transaction.

To support sales and business analysis, a new `Revenue` feature was created using:

**Revenue = Quantity × UnitPrice**

After creating the feature, its basic statistics were examined and revenue was analyzed to understand:

* Which products generate the most revenue
* How revenue changes over time
* Which months generate higher sales revenue
* Whether revenue patterns are related to sales quantity and unit price

The `Revenue` feature will be used in further sales and business analysis.


In [55]:
sales_df["Revenue"] = sales_df["Quantity"] * sales_df["UnitPrice"]

In [56]:
sales_df["Revenue"].describe()

count    524878.000000
mean         20.275399
std         271.693566
min           0.001000
25%           3.900000
50%           9.920000
75%          17.700000
max      168469.600000
Name: Revenue, dtype: float64

In [57]:
sales_df["Revenue"].head(10)

0    15.30
1    20.34
2    22.00
3    20.34
4    20.34
5    15.30
6    25.50
7    11.10
8    11.10
9    54.08
Name: Revenue, dtype: float64

In [58]:
sales_df.groupby("Description")["Revenue"].sum().sort_values(ascending=False).head(10)

Description
DOTCOM POSTAGE                        206248.77
REGENCY CAKESTAND 3 TIER              174156.54
PAPER CRAFT , LITTLE BIRDIE           168469.60
WHITE HANGING HEART T-LIGHT HOLDER    106236.72
PARTY BUNTING                          99445.23
JUMBO BAG RED RETROSPOT                94159.81
MEDIUM CERAMIC TOP STORAGE JAR         81700.92
POSTAGE                                78101.88
Manual                                 77752.82
RABBIT NIGHT LIGHT                     66870.03
Name: Revenue, dtype: float64

In [59]:
sales_df["InvoiceDate"] = pd.to_datetime(sales_df["InvoiceDate"])

In [60]:
sales_df["InvoiceDate"].dtypes

dtype('<M8[us]')

In [61]:
sales_df['InvoiceDate'].min()

Timestamp('2010-12-01 08:26:00')

In [62]:
sales_df['InvoiceDate'].max()

Timestamp('2011-12-09 12:50:00')

### 📅 Feature 2 - Time-Based Analysis

The `InvoiceDate` column contains the date and time of each transaction.
Time-based analysis helps identify **sales trends, order patterns, and periods of higher or lower business activity**.

In this feature, we will investigate:

* Monthly revenue
* Monthly quantity sold
* Monthly number of unique orders
* Order activity across different days of the week


In [63]:
sales_df['InvoiceDate'].dt.month.value_counts().sort_index()

InvoiceDate
1     34060
2     26882
3     35497
4     28882
5     35917
6     35716
7     38395
8     34264
9     48900
10    58537
11    82004
12    65824
Name: count, dtype: int64

In [69]:
sales_df = clean_df[(clean_df["Quantity"] > 0) & (clean_df["UnitPrice"] > 0)].copy()

In [68]:
sales_df["InvoiceDate"].head()

0   2010-12-01 08:26:00
1   2010-12-01 08:26:00
2   2010-12-01 08:26:00
3   2010-12-01 08:26:00
4   2010-12-01 08:26:00
Name: InvoiceDate, dtype: datetime64[us]

In [70]:
sales_df["InvoiceDate"].dtype

dtype('<M8[us]')

In [71]:
sales_df["InvoiceDate"].dt.month.value_counts().sort_index()

InvoiceDate
1     34060
2     26882
3     35497
4     28882
5     35917
6     35716
7     38395
8     34264
9     48900
10    58537
11    82004
12    65824
Name: count, dtype: int64

In [74]:
sales_df["Revenue"] = sales_df["Quantity"] * sales_df["UnitPrice"]

In [77]:
# finding month's Revenue 
sales_df.groupby(sales_df["InvoiceDate"].dt.month)["Revenue"].sum()

InvoiceDate
1      689811.610
2      522545.560
3      716215.260
4      536968.491
5      769296.610
6      760547.010
7      718076.121
8      757841.380
9     1056435.192
10    1151263.730
11    1503866.780
12    1459243.060
Name: Revenue, dtype: float64

In [79]:
# sorting values according to highest sales
sales_df.groupby(sales_df["InvoiceDate"].dt.month)["Revenue"].sum().sort_values(ascending=False)

InvoiceDate
11    1503866.780
12    1459243.060
10    1151263.730
9     1056435.192
5      769296.610
6      760547.010
8      757841.380
7      718076.121
3      716215.260
1      689811.610
4      536968.491
2      522545.560
Name: Revenue, dtype: float64

In [83]:
# November Month's Investigation
sales_df.groupby(sales_df["InvoiceDate"].dt.month)["Quantity"].sum().sort_values(ascending=False)

InvoiceDate
11    751377
12    671631
10    621029
9     569573
8     421020
7     399693
5     395001
6     388511
1     387099
3     376599
4     307953
2     282934
Name: Quantity, dtype: int64

In [85]:
sales_df.groupby( sales_df["InvoiceDate"].dt.month)["UnitPrice"].mean().sort_values(ascending=False)

InvoiceDate
12    4.551426
5     4.170440
3     4.082854
8     4.024723
2     4.002696
4     3.986188
6     3.907644
1     3.787057
10    3.786436
7     3.712157
11    3.625740
9     3.555724
Name: UnitPrice, dtype: float64

In [ ]:
sales_df.groupby( sales_df["InvoiceDate"].dt.month)["InvoiceNo"].nunique()

InvoiceDate
1     1086
2     1100
3     1454
4     1246
5     1681
6     1533
7     1475
8     1361
9     1837
10    2040
11    2769
12    2378
Name: InvoiceNo, dtype: int64

In [108]:
#  sorting values
sales_df.groupby( sales_df["InvoiceDate"].dt.month)["InvoiceNo"].nunique().sort_values(ascending=False)

InvoiceDate
11    2769
12    2378
10    2040
9     1837
5     1681
6     1533
7     1475
3     1454
8     1361
4     1246
2     1100
1     1086
Name: InvoiceNo, dtype: int64

In [113]:
sales_df["DaysofWeek"]=sales_df["InvoiceDate"].dt.day_name()

In [115]:
sales_df["DaysofWeek"].head()

0    Wednesday
1    Wednesday
2    Wednesday
3    Wednesday
4    Wednesday
Name: DaysofWeek, dtype: str

In [118]:
sales_df.groupby("DaysofWeek")["InvoiceNo"].nunique().sort_values(ascending=False)

DaysofWeek
Thursday     4246
Wednesday    3690
Tuesday      3554
Friday       3140
Monday       3126
Sunday       2204
Name: InvoiceNo, dtype: int64

In [119]:
sales_df["DaysofWeek"].unique()


<StringArray>
['Wednesday', 'Thursday', 'Friday', 'Sunday', 'Monday', 'Tuesday']
Length: 6, dtype: str

### 💵 Feature 3 — Price-Based Analysis

The `UnitPrice` column represents the selling price of one unit of a product. Price-based analysis helps us understand how product pricing relates to sales and revenue.

We will investigate:

* The average selling price of products
* How prices vary across products
* The relationship between price, quantity sold, and revenue
* Whether higher revenue is associated with higher selling prices

From the previous time-based analysis, we observed that **November had the highest revenue and the highest quantity sold, but its average `UnitPrice` was not the highest**. This suggests that November's high revenue may have been driven more by **sales volume and order activity than by higher selling prices**.

The `UnitPrice` feature will therefore be analyzed further to understand its role in overall sales performance.


In [125]:
sales_df.groupby("Description")["UnitPrice"].mean()

Description
 4 PURPLE FLOCK DINNER CANDLES       2.450513
 50'S CHRISTMAS GIFT BAG LARGE       1.426589
 DOLLY GIRL BEAKER                   1.506420
 I LOVE LONDON MINI BACKPACK         4.616667
 I LOVE LONDON MINI RUCKSACK         4.150000
                                      ...    
ZINC T-LIGHT HOLDER STARS SMALL      0.943673
ZINC TOP  2 DOOR WOODEN SHELF       21.094167
ZINC WILLIE WINKIE  CANDLE STICK     1.090849
ZINC WIRE KITCHEN ORGANISER          9.929375
ZINC WIRE SWEETHEART LETTER TRAY     3.976522
Name: UnitPrice, Length: 4026, dtype: float64

In [126]:
sales_df.groupby("Description")["UnitPrice"].mean().sort_values(ascending=False).head(10)

Description
Adjust bad debt                       11062.060000
AMAZON FEE                             6880.545000
PICNIC BASKET WICKER 60 PIECES          649.500000
DOTCOM POSTAGE                          292.137068
Manual                                  233.757981
RUSTIC  SEVENTEEN DRAWER SIDEBOARD      158.076923
REGENCY MIRROR WITH SHUTTERS            156.428571
VINTAGE BLUE KITCHEN CABINET            146.750000
VINTAGE RED KITCHEN CABINET             143.421053
CHEST NATURAL WOOD 20 DRAWERS           118.076923
Name: UnitPrice, dtype: float64

In [136]:
price_counts = sales_df.groupby("Description").size()

In [137]:
price_counts.loc[["Adjust bad debt" ,"AMAZON FEE" ,"DOTCOM POSTAGE" ,"Manual"]]

Description
Adjust bad debt      1
AMAZON FEE           2
DOTCOM POSTAGE     706
Manual             317
dtype: int64

In [143]:
revenue_by_description = sales_df.groupby("Description")["Revenue"].sum()

In [144]:
revenue_by_description.loc[ ["Adjust bad debt" ,"AMAZON FEE" ,"DOTCOM POSTAGE" ,"Manual"] ]

Description
Adjust bad debt     11062.06
AMAZON FEE          13761.09
DOTCOM POSTAGE     206248.77
Manual              77752.82
Name: Revenue, dtype: float64